demos/https/iot_demo_https_s3_upload_async.c

```diff
@@ -0,0 +1,538 @@
+    reqConfig.userBuffer.bufferLen = sizeof( _pReqUserBuffer );
+    reqConfig.isAsync = true;
+    reqConfig.u.pAsyncInfo = &asyncInfo;
+
+    respConfig.userBuffer.pBuffer = _pRespUserBuffer;
+    respConfig.userBuffer.bufferLen = sizeof( _pRespUserBuffer );
+    respConfig.pSyncInfo = NULL;
+
+    /* Create the semaphore for waiting for the whole upload response to finish. */
+    uploadFinishedSemCreated = IotSemaphore_Create( &( _uploadFinishedSem ), 0 /* Initial value. */, 1 /* Max value. */ );
+
+    if( uploadFinishedSemCreated == false )
+    {
+        IotLogError( "Failed to create a semaphore to wait for the response to finish." );
+        IOT_SET_AND_GOTO_CLEANUP( EXIT_FAILURE );
+    }
+
+    /* Initialize the HTTPS library. */
+    httpsClientStatus = IotHttpsClient_Init();
+
+    if( httpsClientStatus != IOT_HTTPS_OK )
+    {
+        IotLogError( "An error occurred initializing the HTTPS library. Error code: %d", httpsClientStatus );
+        IOT_SET_AND_GOTO_CLEANUP( EXIT_FAILURE );
+    }
+
+    /* Initialize the request to initialize internal request context, write the HTTP request first line to the
+     * _pReqUserBuffer header space, and write headers User-Agent and Host. */
+    httpsClientStatus = IotHttpsClient_InitializeRequest( &reqHandle, &reqConfig );
+
+    if( httpsClientStatus != IOT_HTTPS_OK )
+    {
+        IotLogError( "An error occurred in IotHttpsClient_InitializeRequest() with error code: %d", httpsClientStatus );
+        IOT_SET_AND_GOTO_CLEANUP( EXIT_FAILURE );
+    }
+
+    /* Connect to S3. */
+    for( connIndex = 1; connIndex <= IOT_DEMO_HTTPS_CONNECTION_NUM_RETRY; connIndex++ )
+    {
+        httpsClientStatus = IotHttpsClient_Connect( &connHandle, &connConfig );
+
+        if( ( httpsClientStatus == IOT_HTTPS_CONNECTION_ERROR ) &&
+            ( connIndex < IOT_DEMO_HTTPS_CONNECTION_NUM_RETRY ) )
+        {
+            IotLogError( "Failed to connect to the S3 server, retrying after %d ms.",
+                         IOT_DEMO_HTTPS_CONNECTION_RETRY_WAIT_MS );
+            IotClock_SleepMs( IOT_DEMO_HTTPS_CONNECTION_RETRY_WAIT_MS );
+            continue;
+        }
+        else
+        {
+            break;
+        }
+    }
+
+    if( httpsClientStatus != IOT_HTTPS_OK )
+    {
+        IotLogError( "Failed to connect to the S3 server. Error code: %d.", httpsClientStatus );
+        IOT_SET_AND_GOTO_CLEANUP( EXIT_FAILURE );
+    }
+
+    /* Send the request and receive the response asynchronously. This will schedule the async request. We
+     * will return immediately after scheduling. */
+    httpsClientStatus = IotHttpsClient_SendAsync( connHandle, reqHandle, &respHandle, &respConfig );
+
+    if( httpsClientStatus != IOT_HTTPS_OK )
+    {
+        IotLogError( "Failed to send the request asynchronously with error code: %d", httpsClientStatus );
+        /* If we failed to schedule an async request then this is an error and we should exit the loop. */
```

Comment:
Is IOT_SET_AND_GOTO_CLEANUP( EXIT_FAILURE ) needed here ?

Reply:
Yes thanks for catching